web3/decoding.nim

```diff
@@ -0,0 +1,416 @@
+    stew/[byteutils, endians2, assign2],
+    serialization,
+    ./eth_api_types,
+    ./abi_utils
+
+from ./abi_serialization import AbiReader
+
+{.push raises: [].}
+
+export abi_serialization
+
+type
+  AbiDecoder* = object
+    input: InputStream
+  UInt = SomeUnsignedInt | StUint
+
+template basetype(Range: type range): untyped =
+  when Range isnot SomeUnsignedInt: {.error: "only uint ranges supported".}
+  elif sizeof(Range) == sizeof(uint8): uint8
+  elif sizeof(Range) == sizeof(uint16): uint16
+  elif sizeof(Range) == sizeof(uint32): uint32
+  elif sizeof(Range) == sizeof(uint64): uint64
+  else:
+    {.error "unsupported range type".}
+
+proc finish(decoder: var AbiDecoder) {.raises: [SerializationError].} =
+  try:
+    if decoder.input.readable:
+      raise newException(SerializationError, "unread trailing bytes found")
+  except IOError as e:
+    raise newException(SerializationError, "Failed to finish decoding: " & e.msg)
+
+proc read(decoder: var AbiDecoder, size = abiSlotSize): seq[byte] {.raises: [SerializationError].} =
+  var buf = newSeq[byte]((size + 31) div 32 * 32)
+
+  try:
+    if not decoder.input.readInto(buf):
+      raise newException(SerializationError, "reading past end of bytes")
+  except IOError as e:
+    raise newException(SerializationError, "Failed to read bytes: " & e.msg)
+
+  return buf
+
+proc readAll(decoder: var AbiDecoder) : seq[byte] {.raises: [SerializationError].} =
+  try:
+    while decoder.input.readable:
+      result.add decoder.input.read
+    return result
+  except IOError as e:
+    raise newException(SerializationError, "Failed to read bytes: " & e.msg)
+
+template checkLeftPadding(buf: openArray[byte], padding: int, expected: uint8) =
+  for i in 0 ..< padding:
+    if buf[i] != expected:
+      raise newException(SerializationError, "invalid padding found")
+
+template checkRightPadding(buf: openArray[byte], paddingStart: int, paddingEnd: int) =
+  for i in paddingStart ..< paddingEnd:
+    if buf[i] != 0x00'u8:
+      raise newException(SerializationError, "invalid padding found")
+
+proc decode(decoder: var AbiDecoder, T: type UInt): T {.raises: [SerializationError].} =
+  var buf = decoder.read(sizeof(T))
+
+  let padding = abiSlotSize - sizeof(T)
+  checkLeftPadding(buf, padding, 0x00'u8)
+
+  return T.fromBytesBE(buf[padding ..< abiSlotSize])
+
+proc decode(decoder: var AbiDecoder, T: type StInt): T {.raises: [SerializationError].} =
+  var buf = decoder.read(sizeof(T))
+
+  let padding = abiSlotSize - sizeof(T)
+  let value = T.fromBytesBE(buf[padding ..< abiSlotSize])
+
+  let b = if value.isNegative: 0xFF'u8 else: 0x00'u8
+  checkLeftPadding(buf, padding, b)
+
+  return value
+
+proc decode(decoder: var AbiDecoder, T: type SomeSignedInt): T {.raises: [SerializationError].} =
+  var buf = decoder.read(sizeof(T))
+
+  let padding = abiSlotSize - sizeof(T)
+  let unsigned = T.toUnsigned.fromBytesBE(buf[padding ..< abiSlotSize])
+  let value = cast[T](unsigned)
```

Comment:
Can we not use a type conversion here instead of a cast?

```suggestion
  let value = T(unsigned)
```

Reply:
On second thought, I'm not sure this adds any checking since the operation is on unsigned integers: https://nim-lang.org/docs/manual.html#statements-and-expressions-type-conversions

Reply:
regarding `cast` vs `T(..)`, the latter will raise a Defect when going from unsigned to signed (but no defect in the other direction) - ie signed values are always range-checked - we should also consider whether to a) allow signed integer decoding, and b) what to do when the conversion overflows, the options being to raise a SerializationError or indeed cast - cast might probably be surprising to nim users which leaves raising or removing the overload